sftp/sftp_test.go

```diff
@@ -0,0 +1,274 @@
+				AuthMethod: "keyAuth",
+				PrivateKey: string(privateKey),
+			},
+			expectedError: nil,
+		},
+		{
+			description: "WithUnsupportedAuthMethod",
+			config: &SSHConfig{
+				HostName:   "HostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "invalidAuth",
+				PrivateKey: "somePrivateKey",
+			},
+			expectedError: fmt.Errorf("unsupported authentication method"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			sshConfig, err := sshClientConfig(tc.config)
+			if tc.expectedError != nil {
+				assert.EqualError(t, tc.expectedError, err.Error())
+				assert.Nil(t, sshConfig)
+			} else {
+				assert.Nil(t, err)
+				assert.NotNil(t, sshConfig)
+			}
+		})
+	}
+}
+
+func TestMain(t *testing.T) {
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	// Let's setup the SSH server
+	publicKeyPath, err := filepath.Abs("testdata/ssh/test_key.pub")
+	require.NoError(t, err)
+	sshServer, err := sshserver.Setup(pool, t,
+		sshserver.WithPublicKeyPath(publicKeyPath),
+		sshserver.WithCredentials("linuxserver.io", ""),
+	)
+	require.NoError(t, err)
+	sshServerHost := fmt.Sprintf("localhost:%d", sshServer.Port)
+	t.Logf("SSH server is listening on %s", sshServerHost)
+
+	// Read private key
+	privateKey, err := os.ReadFile("testdata/ssh/test_key")
+	require.NoError(t, err)
+
+	// Setup ssh client
+	hostname, portStr, err := net.SplitHostPort(sshServerHost)
+	require.NoError(t, err)
+	port, err := strconv.Atoi(portStr)
+	require.NoError(t, err)
+	sshClient, err := NewSSHClient(&SSHConfig{
+		User:        "linuxserver.io",
+		HostName:    hostname,
+		Port:        port,
+		AuthMethod:  "keyAuth",
+		PrivateKey:  string(privateKey),
+		DialTimeout: 10 * time.Second,
+	})
+	require.NoError(t, err)
+
+	// Create session
+	session, err := sshClient.NewSession()
+	require.NoError(t, err)
+	defer session.Close()
+
+	remoteDir := filepath.Join("/tmp", "remote")
+	err = session.Run(fmt.Sprintf("mkdir -p %s", remoteDir))
+	require.NoError(t, err)
+
+	sftpManger, err := NewFileManager(sshClient)
+	require.NoError(t, err)
+
+	// Create local and remote directories within the temporary directory
+	localDir := filepath.Join(t.TempDir(), "local")
+
+	err = os.MkdirAll(localDir, 0o755)
+	require.NoError(t, err)
+
+	// Set up local and remote file paths within their respective directories
+	localFilePath := filepath.Join(localDir, "test_file.json")
+	remoteFilePath := filepath.Join(remoteDir, "test_file.json")
+
+	// Create local file and write data to it
+	localFile, err := os.Create(localFilePath)
+	require.NoError(t, err)
+	defer func() { _ = localFile.Close() }()
+	data := []byte(`{"foo": "bar"}`)
+	err = os.WriteFile(localFilePath, data, 0o644)
+	require.NoError(t, err)
+
+	err = sftpManger.Upload(localFilePath, remoteDir)
+	require.NoError(t, err)
+
+	err = sftpManger.Download(remoteFilePath, localDir)
+	require.NoError(t, err)
+
+	err = sftpManger.Delete(remoteFilePath)
```

Comment:
same here

Reply:
In `Delete` if are unable to delete then it is returning an error. We are already checking if Delete is throwing any error or not.
`require.NoError(t, err) `

Reply:
See [here](https://github.com/rudderlabs/rudder-go-kit/pull/393#discussion_r1551096211). You might get a false positive from `Delete` like it doesn't return an error but it didn't delete. The point of an integration test is to go beyond that and check the actual side effect.